clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -101,88 +101,118 @@ public void configure(Map<String, ?> configs) throws KafkaException {
+                if (!CertificateEntries.create(sslEngineFactory.keystore()).equals(
+                        CertificateEntries.create(newSslEngineFactory.keystore()))) {
                     throw new ConfigException("Keystore DistinguishedName or SubjectAltNames do not match");
                 }
             }
-            if (sslEngineBuilder.truststore() == null && newSslEngineBuilder.truststore() != null) {
+            if (sslEngineFactory.truststore() == null && newSslEngineFactory.truststore() != null) {
                 throw new ConfigException("Cannot add SSL truststore to an existing listener for which no " +
                         "truststore was configured.");
             }
             if (keystoreVerifiableUsingTruststore) {
-                if (sslEngineBuilder.truststore() != null || sslEngineBuilder.keystore() != null) {
-                    SslEngineValidator.validate(sslEngineBuilder, newSslEngineBuilder);
+                if (sslEngineFactory.truststore() != null || sslEngineFactory.keystore() != null) {
+                    SslEngineValidator.validate(sslEngineFactory, newSslEngineFactory);
                 }
             }
-            return newSslEngineBuilder;
+            return newSslEngineFactory;
         } catch (Exception e) {
             log.debug("Validation of dynamic config update of SSLFactory failed.", e);
             throw new ConfigException("Validation of dynamic config update of SSLFactory failed: " + e);
         }
     }
 
     public SSLEngine createSslEngine(String peerHost, int peerPort) {
-        if (sslEngineBuilder == null) {
+        if (sslEngineFactory == null) {
             throw new IllegalStateException("SslFactory has not been configured.");
         }
-        return sslEngineBuilder.createSslEngine(mode, peerHost, peerPort, endpointIdentification);
+        if (mode == Mode.SERVER) {
+            return sslEngineFactory.createServerSslEngine(peerHost, peerPort);
+        } else {
+            return sslEngineFactory.createClientSslEngine(peerHost, peerPort, endpointIdentification);
+        }
     }
 
-    @Deprecated
-    public SSLContext sslContext() {
-        return sslEngineBuilder.sslContext();
+    public SslEngineFactory sslEngineFactory() {
+        return sslEngineFactory;
     }
 
-    public SslEngineBuilder sslEngineBuilder() {
-        return sslEngineBuilder;
+    /**
+     * Copy all entries for which the key starts with {@link SslConfigs#SSL_CONFIG_PREFIX} from one map into another.
+     *
+     * @param destMap   The map to copy entries into.
+     * @param srcMap    The map to copy entries from.
+     */
+    private static void copyAllSslConfigs(Map<String, Object> destMap,
+                                              Map<String, ? extends Object> srcMap) {
+        for (Map.Entry<String, ?> entry : srcMap.entrySet()) {
+            if (entry.getKey().startsWith(SslConfigs.SSL_CONFIG_PREFIX)) {
```

Comment:
It would be safer to just give all configs rather than filter out by SSL prefix. We don't want to build any assumptions about custom config names. We give all configs to other plugins anyway.